File: client-python/ui/banner.py
import os
import platform


class Banner:
    LARGURA = 60
    NOME = "BancoAPI"
    VERSAO = "v3.0 GO · Json · API"

    SEPARADOR = "─" * LARGURA
    SEPARADOR_DUPLO = "· " * (LARGURA // 2)

    # ── ANSI RESET ──────────────────────────────────────────────────
    RESET = "\u001B[0m"
    BOLD = "\u001B[1m"

    # ── Paleta para terminal claro ─────────────────────────────────
    CINZA = "\u001B[38;5;245m"
    AZUL = "\u001B[38;5;33m"
    VERDE = "\u001B[38;5;35m"
    VERMELHO = "\u001B[38;5;160m"
    AMARELO = "\u001B[38;5;136m"
    CIANO = "\u001B[38;5;37m"

    B_AZUL = BOLD + AZUL
    B_VERDE = BOLD + VERDE
    B_VERM = BOLD + VERMELHO
    B_AMAR = BOLD + AMARELO
    B_CIANO = BOLD + CIANO
    B_CINZA = BOLD + CINZA

    # ── Logo ────────────────────────────────────────────────────────
    LOGO = [
        " ██████╗  █████╗ ███╗   ██╗ ██████╗ ██████╗ ",
        " ██╔══██╗██╔══██╗████╗  ██║██╔════╝██╔═══██╗",
        " ██████╔╝███████║██╔██╗ ██║██║     ██║   ██║",
        " ██╔══██╗██╔══██║██║╚██╗██║██║     ██║   ██║",
        " ██████╔╝██║  ██║██║ ╚████║╚██████╗╚██████╔╝",
        " ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝ ╚═════╝ "
    ]

    # ── Layout ─────────────────────────────────────────────────────

    @staticmethod
    def margem():
        return "  "

    @staticmethod
    def limpar():
        try:
            sistema = platform.system().lower()
            if "win" in sistema:
                os.system("cls")
            else:
                os.system("clear")
        except Exception:
            print("\033[H\033[2J", end="")

    @staticmethod
    def exibir_cabecalho(subtitulo=None):
        m = Banner.margem()
        print()

        meio = len(Banner.LOGO) // 2

        for i in range(len(Banner.LOGO)):
            sufixo = f"   {Banner.B_CINZA}{Banner.NOME}{Banner.RESET}" if i == meio else ""
            print(f"{m}{Banner.B_AZUL}{Banner.LOGO[i]}{Banner.RESET}{sufixo}")

        print(f"{m}{Banner.CINZA}{Banner.pad_esquerda(Banner.VERSAO, Banner.LARGURA)}{Banner.RESET}")
        print(f"{m}{Banner.CINZA}{Banner.SEPARADOR_DUPLO}{Banner.RESET}")

        if subtitulo and subtitulo.strip():
            print(f"{m}{Banner.B_AMAR}{Banner.centralizar(subtitulo, Banner.LARGURA)}{Banner.RESET}")

        print()

    @staticmethod
    def titulo_secao(texto):
        m = Banner.margem()
        li = Banner.LARGURA - 2

        print(f"{m}{Banner.CINZA}┌{'─' * li}┐{Banner.RESET}")
        print(
            f"{m}{Banner.CINZA}│{Banner.RESET}{Banner.B_CINZA}{Banner.centralizar(texto, li)}{Banner.RESET}{Banner.CINZA}│{Banner.RESET}")
        print(f"{m}{Banner.CINZA}└{'─' * li}┘{Banner.RESET}")
        print()

    @staticmethod
    def label_secao(texto):
        m = Banner.margem()
        print(f"{m}{Banner.B_AZUL}▸ {texto}{Banner.RESET}")
        print(f"{m}{Banner.CINZA}{Banner.SEPARADOR}{Banner.RESET}")

    @staticmethod
    def campo(label, valor, largura_label=16):
        m = Banner.margem()
        rotulo = f"{Banner.CINZA}{label:<{largura_label}}{Banner.RESET}"
        print(f"{m}  {rotulo}{valor}")

    # ── Feedback ────────────────────────────────────────────────────

    @staticmethod
    def sucesso(msg):
        print()
        print(f"{Banner.margem()}{Banner.B_VERDE}  ✔  {msg}{Banner.RESET}")

    @staticmethod
    def erro(msg):
        print()
        print(f"{Banner.margem()}{Banner.B_VERM}  ✖  {msg}{Banner.RESET}")

    @staticmethod
    def info(msg):
        print(f"{Banner.margem()}{Banner.B_CIANO}  ℹ  {msg}{Banner.RESET}")

    # ── Espaçamento ─────────────────────────────────────────────────

    @staticmethod
    def espaco():
        print()

    @staticmethod
    def linha_separadora():
        print(f"{Banner.margem()}{Banner.CINZA}{Banner.SEPARADOR}{Banner.RESET}")

    @staticmethod
    def aguarde_enter():
        print(f"{Banner.margem()}{Banner.CINZA}  Pressione ENTER para continuar...{Banner.RESET}", end="")
        input()

    # ── Formatação monetária ────────────────────────────────────────

    @staticmethod
    def formatar_dinheiro(valor):
        return f"R$ {valor:.2f}"

    @staticmethod
    def valor_positivo(valor):
        return f"{Banner.B_VERDE}{Banner.formatar_dinheiro(valor)}{Banner.RESET}"

    @staticmethod
    def valor_negativo(valor):
        return f"{Banner.B_VERM}{Banner.formatar_dinheiro(valor)}{Banner.RESET}"

    # ── Badges ──────────────────────────────────────────────────────

    @staticmethod
    def badge(texto):
        return f"{Banner.B_CIANO} {texto} {Banner.RESET}"

    @staticmethod
    def badge_verde(texto):
        return f"{Banner.B_VERDE} {texto} {Banner.RESET}"

    @staticmethod
    def badge_azul(texto):
        return f"{Banner.B_AZUL} {texto} {Banner.RESET}"

    # ── Helpers de cor ──────────────────────────────────────────────

    @staticmethod
    def ciano(s):
        return f"{Banner.B_CIANO}{s}{Banner.RESET}"

    @staticmethod
    def verde(s):
        return f"{Banner.B_VERDE}{s}{Banner.RESET}"

    @staticmethod
    def vermelho(s):
        return f"{Banner.B_VERM}{s}{Banner.RESET}"

    @staticmethod
    def amarelo(s):
        return f"{Banner.B_AMAR}{s}{Banner.RESET}"

    @staticmethod
    def cinza(s):
        return f"{Banner.CINZA}{s}{Banner.RESET}"

    @staticmethod
    def branco(s):
        return f"{Banner.B_CINZA}{s}{Banner.RESET}"

    @staticmethod
    def bold(s):
        return f"{Banner.BOLD}{s}{Banner.RESET}"

    # ── Helpers de layout ───────────────────────────────────────────

    @staticmethod
    def centralizar(texto, largura):
        if len(texto) >= largura:
            return texto
        pad = (largura - len(texto)) // 2
        return " " * pad + texto + " " * (largura - len(texto) - pad)

    @staticmethod
    def pad_esquerda(texto, largura):
        if len(texto) >= largura:
            return texto
        return " " * (largura - len(texto)) + texto

    # ── Prompt ──────────────────────────────────────────────────────

    @staticmethod
    def prompt(label):
        print(f"{Banner.margem()}  {Banner.B_AZUL}{label}{Banner.RESET} ", end="")

File: client-java/src/main/java/com/example/client_java/ui/Banner.java
package com.example.client_java.ui;

public class Banner {

    public static final int LARGURA = 60;

    public static final String NOME = "Banco RMI";
    public static final String VERSAO = "v2.0 Java · Protobuf · RMI";

    public static final String SEPARADOR = "─".repeat(LARGURA);
    public static final String SEPARADOR_DUPLO = "· ".repeat(LARGURA / 2);

    // ── ANSI RESET ──────────────────────────────────────────────────
    private static final String RESET = "\u001B[0m";
    private static final String BOLD = "\u001B[1m";

    // ── Paleta para terminal claro ─────────────────────────────────
    private static final String CINZA = "\u001B[38;5;245m";
    private static final String AZUL = "\u001B[38;5;33m";
    private static final String VERDE = "\u001B[38;5;35m";
    private static final String VERMELHO = "\u001B[38;5;160m";
    private static final String AMARELO = "\u001B[38;5;136m";
    private static final String CIANO = "\u001B[38;5;37m";

    private static final String B_AZUL = BOLD + AZUL;
    private static final String B_VERDE = BOLD + VERDE;
    private static final String B_VERM = BOLD + VERMELHO;
    private static final String B_AMAR = BOLD + AMARELO;
    private static final String B_CIANO = BOLD + CIANO;
    private static final String B_CINZA = BOLD + CINZA;

    // ── Logo ────────────────────────────────────────────────────────
    private static final String[] LOGO = {
            " ██████╗  █████╗ ███╗   ██╗ ██████╗ ██████╗ ",
            " ██╔══██╗██╔══██╗████╗  ██║██╔════╝██╔═══██╗",
            " ██████╔╝███████║██╔██╗ ██║██║     ██║   ██║",
            " ██╔══██╗██╔══██║██║╚██╗██║██║     ██║   ██║",
            " ██████╔╝██║  ██║██║ ╚████║╚██████╗╚██████╔╝",
            " ╚═════╝ ╚═╝  ╚═╝╚═╝  ╚═══╝ ╚═════╝ ╚═════╝ "
    };

    // ── Layout ─────────────────────────────────────────────────────

    public static String margem() {
        return "  ";
    }

    public static void limpar() {
        try {
            String os = System.getProperty("os.name").toLowerCase();

            ProcessBuilder pb = os.contains("win")
                    ? new ProcessBuilder("cmd", "/c", "cls")
                    : new ProcessBuilder("clear");

            pb.inheritIO().start().waitFor();

        } catch (Exception e) {
            System.out.print("\033[H\033[2J");
            System.out.flush();
        }
    }

    public static void exibirCabecalho() {
        exibirCabecalho(null);
    }

    public static void exibirCabecalho(String subtitulo) {

        String m = margem();

        System.out.println();

        int meio = LOGO.length / 2;

        for (int i = 0; i < LOGO.length; i++) {

            String sufixo = (i == meio)
                    ? "   " + B_CINZA + NOME + RESET
                    : "";

            System.out.println(
                    m + B_AZUL + LOGO[i] + RESET + sufixo
            );
        }

        System.out.println(
                m + CINZA + padEsquerda(VERSAO, LARGURA) + RESET
        );

        System.out.println(
                m + CINZA + SEPARADOR_DUPLO + RESET
        );

        if (subtitulo != null && !subtitulo.isBlank()) {
            System.out.println(
                    m + B_AMAR + centralizar(subtitulo, LARGURA) + RESET
            );
        }

        System.out.println();
    }

    public static void tituloSecao(String texto) {

        String m = margem();
        int li = LARGURA - 2;

        System.out.println(
                m + CINZA + "┌" + "─".repeat(li) + "┐" + RESET
        );

        System.out.println(
                m + CINZA + "│"
                        + RESET
                        + B_CINZA
                        + centralizar(texto, li)
                        + RESET
                        + CINZA
                        + "│"
                        + RESET
        );

        System.out.println(
                m + CINZA + "└" + "─".repeat(li) + "┘" + RESET
        );

        System.out.println();
    }

    public static void labelSecao(String texto) {

        String m = margem();

        System.out.println(
                m + B_AZUL + "▸ " + texto + RESET
        );

        System.out.println(
                m + CINZA + SEPARADOR + RESET
        );
    }

    public static void campo(String label, String valor) {
        campo(label, valor, 16);
    }

    public static void campo(String label, String valor, int larguraLabel) {

        String m = margem();

        String rotulo = CINZA
                + String.format("%-" + larguraLabel + "s", label)
                + RESET;

        System.out.println(
                m + "  " + rotulo + valor
        );
    }

    // ── Feedback ────────────────────────────────────────────────────

    public static void sucesso(String msg) {

        System.out.println();

        System.out.println(
                margem()
                        + B_VERDE
                        + "  ✔  "
                        + msg
                        + RESET
        );
    }

    public static void erro(String msg) {

        System.out.println();

        System.out.println(
                margem()
                        + B_VERM
                        + "  ✖  "
                        + msg
                        + RESET
        );
    }

    public static void info(String msg) {

        System.out.println(
                margem()
                        + B_CIANO
                        + "  ℹ  "
                        + msg
                        + RESET
        );
    }

    // ── Espaçamento ─────────────────────────────────────────────────

    public static void espaco() {
        System.out.println();
    }

    public static void linhaSeparadora() {

        System.out.println(
                margem()
                        + CINZA
                        + SEPARADOR
                        + RESET
        );
    }

    public static void aguardeEnter(java.util.Scanner sc) {

        System.out.print(
                margem()
                        + CINZA
                        + "  Pressione ENTER para continuar..."
                        + RESET
        );

        sc.nextLine();
    }

    // ── Formatação monetária ────────────────────────────────────────

    public static String formatarDinheiro(double valor) {
        return String.format("R$ %.2f", valor);
    }

    public static String valorPositivo(double valor) {
        return B_VERDE + formatarDinheiro(valor) + RESET;
    }

    public static String valorNegativo(double valor) {
        return B_VERM + formatarDinheiro(valor) + RESET;
    }

    // ── Badges ──────────────────────────────────────────────────────

    public static String badge(String texto) {
        return B_CIANO + " " + texto + " " + RESET;
    }

    public static String badgeVerde(String texto) {
        return B_VERDE + " " + texto + " " + RESET;
    }

    public static String badgeAzul(String texto) {
        return B_AZUL + " " + texto + " " + RESET;
    }

    // ── Helpers de cor ──────────────────────────────────────────────

    public static String ciano(String s) {
        return B_CIANO + s + RESET;
    }

    public static String verde(String s) {
        return B_VERDE + s + RESET;
    }

    public static String vermelho(String s) {
        return B_VERM + s + RESET;
    }

    public static String amarelo(String s) {
        return B_AMAR + s + RESET;
    }

    public static String cinza(String s) {
        return CINZA + s + RESET;
    }

    public static String branco(String s) {
        return B_CINZA + s + RESET;
    }

    public static String bold(String s) {
        return BOLD + s + RESET;
    }

    // ── Helpers de layout ───────────────────────────────────────────

    public static String centralizar(String texto, int largura) {

        if (texto.length() >= largura) {
            return texto;
        }

        int pad = (largura - texto.length()) / 2;

        return " ".repeat(pad)
                + texto
                + " ".repeat(largura - texto.length() - pad);
    }

    public static String padEsquerda(String texto, int largura) {

        if (texto.length() >= largura) {
            return texto;
        }

        return " ".repeat(largura - texto.length()) + texto;
    }

    // ── Prompt ──────────────────────────────────────────────────────

    public static void prompt(String label) {

        System.out.print(
                margem()
                        + "  "
                        + B_AZUL
                        + label
                        + RESET
                        + " "
        );
    }
}
